src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+        |> System.Text.Encoding.UTF8.GetString
+      if String.IsNullOrWhiteSpace message then
+        return None
+      else
+        let! result =
+          message
+          |> deserializeClientMessage serializerOptions
+        return Some result
+    }
+
+  let sendMessageViaSocket (jsonSerializerOptions) (socket : WebSocket) (message : ServerMessage) =
+    task {
+      if not (socket.State = WebSocketState.Open) then
+        if logger.IsEnabled(LogLevel.Trace) then
+          logger.LogTrace("Ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+      else
+        let! serializedMessage = message |> serializeServerMessage jsonSerializerOptions
+        let segment =
+          new ArraySegment<byte>(
+            System.Text.Encoding.UTF8.GetBytes(serializedMessage)
+          )
+        if not (socket.State = WebSocketState.Open) then
+          if logger.IsEnabled(LogLevel.Trace) then
+            logger.LogTrace("ignoring message to be sent via socket, since its state is not 'Open', but '{state}'", socket.State)
+        else
+          do! socket.SendAsync(segment, WebSocketMessageType.Text, endOfMessage = true, cancellationToken = new CancellationToken())
+
+        if logger.IsEnabled(LogLevel.Trace) then
+          logger.LogTrace("<- Response: {response}", message)
+    }
+
+  let addClientSubscription (id : SubscriptionId) (jsonSerializerOptions) (socket) (howToSendDataOnNext: SubscriptionId -> Output -> Task<unit>) (streamSource: IObservable<Output>) (subscriptions : SubscriptionsDict)  =
+    let observer = new Reactive.AnonymousObserver<Output>(
+      onNext =
+        (fun theOutput ->
+          theOutput
+          |> howToSendDataOnNext id
+          |> Task.WaitAll
+        ),
+      onError =
+        (fun ex ->
+          logger.LogError("[Error on subscription {id}]: {exceptionstr}", id, ex)
+        ),
+      onCompleted =
+        (fun () ->
+          Complete id
+          |> sendMessageViaSocket jsonSerializerOptions (socket)
+          |> Async.AwaitTask
+          |> Async.RunSynchronously
+          subscriptions
+          |> GraphQLSubscriptionsManagement.removeSubscription(id)
+        )
+    )
+
+    let unsubscriber = streamSource.Subscribe(observer)
+
+    subscriptions
+    |> GraphQLSubscriptionsManagement.addSubscription(id, unsubscriber, (fun _ -> ()))
+
+  let tryToGracefullyCloseSocket (code, message) theSocket =
+    task {
+        if theSocket |> canCloseSocket
+        then
+          do! theSocket.CloseAsync(code, message, new CancellationToken())
+        else
+          ()
+    }
+
+  let tryToGracefullyCloseSocketWithDefaultBehavior =
+    tryToGracefullyCloseSocket (WebSocketCloseStatus.NormalClosure, "Normal Closure")
+
+  let handleMessages (cancellationToken: CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (root: unit -> 'Root) (pingHandler : PingHandler option) (socket : WebSocket) =
```

Comment:
Maybe here too. Reduce parameters by grouping them into a single object.

Reply:
Hm... here I wouldn't do this because this function is just called once, so I don't see what benefit we'd get by introducing this other layer of code just not to pass the arguments directly.